plugin/evm/vm.go

```diff
@@ -652,6 +680,37 @@ func (vm *VM) buildBlock(context.Context) (snowman.Block, error) {
 	return blk, nil
 }
 
+// checkTransactionPredicates checks the stateful precompile predicates of any of the given
+// transactions that reference a stateful precompile address in their access list.
+// Returns [len(txs), nil] if and only if all referenced predicates are met.
+// Otherwise, returns the index of the first transaction that was invalid and the predicate error.
+func (vm *VM) checkTransactionPredicates(txs types.Transactions, proposerVMBlockCtx *block.Context) (int, error) {
+	precompileConfigs := vm.currentRules().Precompiles
+	for i, tx := range txs {
+		for _, accessTuple := range tx.AccessList() {
+			var (
+				precompileConfig   precompile.StatefulPrecompileConfig
+				isPrecompileAccess bool
+			)
+			if precompileConfig, isPrecompileAccess = precompileConfigs[accessTuple.Address]; !isPrecompileAccess {
+				continue
+			}
+
+			predicate := precompileConfig.Predicate()
+			if predicate == nil {
+				continue
+			}
+
+			if err := predicate(vm.ctx, proposerVMBlockCtx, accessTuple.StorageKeys); err != nil {
+				log.Warn("Transaction predicate verification failed.", "txId", tx.Hash(), "precompileAddress", accessTuple.Address.Hex())
```

Comment:
could we use `Debug` instead? we should avoid emitting warn logs where an end user could cause lots of warn logs by submitting invalid txs.